src/AgentCommerce/HoneyWebhookService.php

```diff
@@ -0,0 +1,175 @@
+<?php declare(strict_types=1);
+/*
+ * (c) shopware AG <info@shopware.com>
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Swag\PayPal\AgentCommerce;
+
+use GuzzleHttp\ClientInterface;
+use GuzzleHttp\Exception\ClientException;
+use Lcobucci\JWT\Encoding\ChainedFormatter;
+use Lcobucci\JWT\Encoding\JoseEncoder;
+use Lcobucci\JWT\Signer\Hmac\Sha256;
+use Lcobucci\JWT\Signer\Key\InMemory;
+use Lcobucci\JWT\Token\Builder;
+use Psr\Log\LoggerInterface;
+use Shopware\Core\Framework\Context;
+use Shopware\Core\Framework\DataAbstractionLayer\EntityRepository;
+use Shopware\Core\Framework\DataAbstractionLayer\Search\Criteria;
+use Shopware\Core\Framework\DataAbstractionLayer\Search\Filter\EqualsFilter;
+use Shopware\Core\Framework\Log\Package;
+use Shopware\Core\System\Country\CountryEntity;
+use Shopware\Core\System\SalesChannel\SalesChannelCollection;
+use Shopware\Core\System\SalesChannel\SalesChannelEntity;
+use Shopware\Core\System\SystemConfig\SystemConfigService;
+use Swag\PayPal\AgentCommerce\Exception\HoneyWebhookExceptions;
+use Swag\PayPal\Setting\Service\CredentialsUtil;
+use Swag\PayPal\Setting\Settings;
+use Swag\PayPal\SwagPayPal;
+use Symfony\Component\Routing\RouterInterface;
+
+/**
+ * @internal
+ */
+#[Package('checkout')]
+class HoneyWebhookService
+{
+    /**
+     * @param EntityRepository<SalesChannelCollection> $salesChannelRepository
+     */
+    public function __construct(
+        private readonly ClientInterface $client,
+        private readonly EntityRepository $salesChannelRepository,
+        private readonly CredentialsUtil $credentialsUtil,
+        private readonly RouterInterface $router,
+        private readonly SystemConfigService $systemConfigService,
+        private readonly LoggerInterface $logger,
+    ) {
+    }
+
+    public function register(string $salesChannelId, Context $context): HoneyWebhookResult
+    {
+        try {
+            $salesChannel = $this->loadSalesChannel($salesChannelId, $context);
+            if (!$salesChannel || !$salesChannel->getActive()) {
+                throw HoneyWebhookExceptions::invalidSalesChannel();
+            }
+
+            $oldToken = $this->systemConfigService->get(Settings::AGENT_COMMERCE_ONBOARDED, $salesChannelId);
+            if (\is_string($oldToken)) {
+                $deregisterResult = $this->webhookCall($oldToken, 'uninstall');
+
+                if (!$deregisterResult->success) {
+                    throw HoneyWebhookExceptions::failedDeregisterWebhook();
+                }
+            }
```

Comment:
Can't we call the deregister function? Removing the token from the system config after successful deregistration should be fine, not?

Reply:
don't catch it :smile: you could also move the sales channel check into the `createToken` function and do:
```php
try {
   $token = $this->createToken($salesChannelId);
} catch ($e) {
    $this->logger->error('... cannot create token', [...]);
    throw $e
}

try {
   $this->deregister();
} catch (...) {}

// register
```
mh indeed, thats a lot of try catches 💢